### PR TITLE
Handle settings when using proxy/load balancer or containerized

### PR DESCRIPTION
**Pull request by toch**
Those changes:
- make sure the build args are set up even when we run `docker build`
- handle passed env vars to set the right ones for Django

this is linked to https://github.com/BLSQ/openhexa/pull/44

### CODE REVIEW

**Comment by qgerome** [comment]
> > WDYT of moving this code in the entry point into the local.py configuration file?
> 
> isn't it useful for the other cases?

It is indeed. Your changes are generic enough to even be put in the `base.py` settings IMO. It would allow us to keep the number of configuration files as minimum as possible. WDYT ?

**Comment by toch** [comment]
I'm going to take care of it.

**Comment by toch** [comment]
Isn't my `PROXY_URL` the same than the existing `BASE_URL`?

File: config/settings/base.py [hunk 1]
```diff
@@ -31,10 +31,96 @@
...
+    SESSION_COOKIE_SECURE = True
+    CSRF_COOKIE_SECURE = True
+    SECURE_SSL_REDIRECT = True
+else:
+    SESSION_COOKIE_SECURE = False
+    CSRF_COOKIE_SECURE = False
+    SECURE_SSL_REDIRECT = False
 
+BASE_HOSTNAME = os.environ.get("BASE_HOSTNAME", "localhost")
+BASE_PORT = os.environ.get("BASE", 8000)
```
Comment (qgerome) [inline config/settings/base.py:56]:
```suggestion
BASE_PORT = os.environ.get("BASE_PORT", 8000)
```

**Review by qgerome — COMMENTED** [review]
Shouldn't we add the new environment variables in the .env.dist ?

**Comment by toch** [comment]
> Shouldn't we add the new environment variables in the .env.dist ?

Yes, thx

**Comment by toch** [comment]
I also need to update the compose manifest file, I'll do that tomorrow

File: .env.dist [hunk 2]
```diff
@@ -32,13 +32,38 @@ POSTGRES_DB=hexa-app
...
+# If not set, it falls back to either PROXY_HOSTNAME_AND_PORT or
+# BASE_HOSTNAME:FRONTEND_PORT
+# NEW_FRONTEND_DOMAIN=http://localhost:3000
+
+# The port number to access the frontend
+FRONTEND_PORT=3000
+# The port number to access the Jupyter hub
+JUPYTERHUB_PORT=8001
+
+# I'd put that directly in the compose manifest file
```
[the remark this answers is not in the document]
Comment (qgerome) [reply .env.dist:59]: I'm interested, what would be the explanation to move this variable to the compose.yaml and not for example "INTERNAL_BASE_URL" or "NEW_FRONTEND_DOMAIN" ?

File: config/settings/base.py [hunk 3]
```diff
@@ -31,10 +31,97 @@
...
 DEBUG = os.environ.get("DEBUG", "false") == "true"
 
-ALLOWED_HOSTS = os.environ.get("ALLOWED_HOSTS", "").split(",")
+# Trust the X_FORWARDED_PROTO header from the proxy or load balancer so Django is aware it is accessed by https
+if "TRUST_FORWARDED_PROTO" in os.environ:
+    SECURE_PROXY_SSL_HEADER = ("HTTP_X_FORWARDED_PROTO", "https")
+
+TLS = os.environ.get("TLS", "false") == "true"
+SCHEME = "https" if TLS else "http"
+# Old version
```
[the remark this answers is not in the document]
Comment (qgerome) [reply config/settings/base.py:40]: Yes I saw, thanks for that, it's gonna be helpful 🙏🏻

File: .env.dist (hunk 2 above)
[the remark this answers is not in the document]
Comment (toch) [reply .env.dist:59]: As `INTERNAL_BASE_URL` has been introduced in the last commits, I admit I didn't think about it. I think I could maybe deduce it based on the context in the `base.py`.

For `NEW_FRONTEND_DOMAIN`, I kept it as you have a setup on GCP where you expose each service on a subdomain. But that can be changed too.

File: config/settings/base.py [hunk 4]
```diff
@@ -80,9 +81,8 @@
     NEW_FRONTEND_DOMAIN = os.environ.get(
         "NEW_FRONTEND_DOMAIN", os.environ.get("PROXY_HOSTNAME_AND_PORT")
     )
-    if not NEW_FRONTEND_DOMAIN.lower().startswith(
-        "http://"
-    ) and not NEW_FRONTEND_DOMAIN.lower().startswith("https://"):
+    if not re.match(r"^https?://", NEW_FRONTEND_DOMAIN, re.IGNORECASE):
```
Comment (toch) [inline config/settings/base.py:84]: For the record, that was my initial implementation. I opted for a simpler implementation than using regexp as it's something that is not well mastered. From a performance point of view, `startswith` is also more efficient, but considering that we can mix two together, we should measure to confirm which is which.

I'm fine with your changes.

**Pull request merged**